tests/functional/test_useragent.py

```diff
@@ -0,0 +1,330 @@
+
+
+@pytest.mark.parametrize(
+    'sess_name, sess_version, sess_extra, cfg_extra',
+    # Produce every combination of User-Agent related config settings other
+    # than Config.user_agent which is never set in this test and
+    # Config.user_agent_appid which is always set in this test.
+    product(
+        ('sess_name', None),
+        ('sess_version', None),
+        ('sess_extra', None),
+        ('cfg_extra', None),
+    ),
+)
+def test_user_agent_includes_appid(
+    sess_name,
+    sess_version,
+    sess_extra,
+    cfg_extra,
+    patched_session,
+    stubbed_list_buckets,
+):
+    # The User-Agent header string should always include the value set in
+    # ``Config.user_agent_appid``, unless ``Config.user_agent`` is also set
+    # (this latter case is covered in ``test_user_agent_from_config_replaces_default``).
+    if sess_name:
+        patched_session.user_agent_name = sess_name
+    if sess_version:
+        patched_session.user_agent_version = sess_version
+    if sess_extra:
+        patched_session.user_agent_extra = sess_extra
+    client_cfg = Config(
+        user_agent=None,
+        user_agent_appid='123456',
+        user_agent_extra=cfg_extra,
+    )
+    client_s3 = patched_session.create_client('s3', config=client_cfg)
+    with uacap_client(client_s3) as cap_client:
+        with stubbed_list_buckets(cap_client):
+            cap_client.list_buckets()
+
+    uafields = cap_client.captured_ua_string.split(' ')
+    assert 'app/123456' in uafields
+
+
+def test_user_agent_long_appid_yields_warning(
+    patched_session, stubbed_list_buckets, caplog
+):
+    # user_agent_appid config values longer than 50 characters should result
+    # in a warning
+    sixtychars = '000000000011111111112222222222333333333344444444445555555555'
+    client_cfg = Config(user_agent_appid=sixtychars)
+    client_s3 = patched_session.create_client('s3', config=client_cfg)
+    with uacap_client(client_s3) as cap_client:
+        with stubbed_list_buckets(cap_client):
+            with caplog.at_level(logging.INFO):
+                cap_client.list_buckets()
+
+    assert (
+        'The configured value for user_agent_appid exceeds the maximum length'
+        in caplog.text
+    )
+
+
+def test_user_agent_appid_gets_sanitized(
+    patched_session, stubbed_list_buckets, caplog
+):
+    # Parentheses and the copyright symbol are not valid characters in the user
+    # agent string
+    badchars = '1234('
+    client_cfg = Config(user_agent_appid=badchars)
+    client_s3 = patched_session.create_client('s3', config=client_cfg)
+    with uacap_client(client_s3) as cap_client:
+        with stubbed_list_buckets(cap_client):
+            with caplog.at_level(logging.INFO):
+                cap_client.list_buckets()
+
+    # given string should be truncated to 50 characters
+    uafields = cap_client.captured_ua_string.split(' ')
+    assert 'app/1234-' in uafields
+
+
+def test_boto3_behavior(patched_session, stubbed_list_buckets):
```

Comment:
nit; We may want to be more specific than behavior.
```suggestion
def test_boto3_user_agent(patched_session, stubbed_list_buckets):
```

Reply:
Done in eb7ded6